can/interfaces/cantact.py

```diff
@@ -0,0 +1,135 @@
+"""
+Interface for CANtact devices from Linklayer Labs
+"""
+
+import time
+import logging
+from unittest.mock import Mock
+
+from can import BusABC, Message
+
+logger = logging.getLogger(__name__)
+
+try:
+    import cantact
+except ImportError:
+    logger.warning(
+        "The CANtact module is not installed. Install it using `python3 -m pip install cantact`"
+    )
+
+
+class CantactBus(BusABC):
+    """CANtact interface"""
+
+    @staticmethod
+    def _detect_available_configs():
+        try:
+            interface = cantact.Interface()
+        except NameError:
+            # couldn't import cantact, so no configurations are available
+            return []
+
+        channels = []
+        for i in range(0, interface.channel_count()):
+            channels.append({"interface": "cantact", "channel": "ch:%d" % i})
+        return channels
+
+    def __init__(
+        self,
+        channel,
+        bitrate=500000,
+        poll_interval=0.01,
+        monitor=False,
+        _testing=False,
+        **kwargs
+    ):
+        """
+        :param int channel:
+            Device number
+        :param int bitrate:
+            Bitrate in bits/s
+        :param bool monitor:
+            If true, operate in listen-only monitoring mode
+        """
+
+        if _testing:
+            self.interface = MockInterface()
+        else:
+            self.interface = cantact.Interface()
+
+        self.channel = int(channel)
+        self.channel_info = "CANtact: ch:%s" % channel
+
+        # configure the interface
+        self.interface.set_bitrate(int(channel), int(bitrate))
+        self.interface.set_enabled(int(channel), True)
+        self.interface.set_monitor(int(channel), monitor)
+        self.interface.start()
+
+        super().__init__(
+            channel=channel, bitrate=bitrate, poll_interval=poll_interval, **kwargs
+        )
+
+    def _recv_internal(self, timeout):
+        frame = self.interface.recv(int(timeout * 1000))
+        if frame is None:
+            # timeout occured
+            return None, False
+
+        msg = Message(
+            arbitration_id=frame["id"],
+            is_extended_id=frame["extended"],
+            timestamp=time.time(),  # Better than nothing...
```

Comment:
Is there no hardware timestamp?

Reply:
This varies by device, but I think it makes sense to move that to the driver and have it determine the timestamp if available (or use system time if not).